src/jit/morph.cpp

```diff
@@ -8804,64 +9249,90 @@ GenTreePtr Compiler::fgMorphCopyBlock(GenTreePtr tree)
     }
     else
     {
-        GenTreePtr destAddr;
-        GenTreePtr srcAddr;
-        GenTreePtr blockSize;
-        bool       isCopyObj;
+        unsigned             blockWidth;
+        bool                 blockWidthIsConst = false;
+        GenTreeLclVarCommon* lclVarTree        = nullptr;
+        GenTreeLclVarCommon* srcLclVarTree     = nullptr;
+        unsigned             destLclNum        = BAD_VAR_NUM;
+        LclVarDsc*           destLclVar        = nullptr;
+        FieldSeqNode*        destFldSeq        = nullptr;
+        bool                 destDoFldAsg      = false;
+        GenTreePtr           destAddr          = nullptr;
+        GenTreePtr           srcAddr           = nullptr;
+        bool                 destOnStack       = false;
+        bool                 hasGCPtrs         = false;
 
-        if (tree->OperGet() == GT_COPYBLK)
-        {
-            GenTreeCpBlk* copyBlkOp = tree->AsCpBlk();
-
-            isCopyObj = false;
-            destAddr  = copyBlkOp->Dest();
-            srcAddr   = copyBlkOp->Source();
-            blockSize = copyBlkOp->Size();
-        }
-        else
-        {
-            GenTreeCpObj* copyObjOp = tree->AsCpObj();
-
-            isCopyObj = true;
-            destAddr  = copyObjOp->Dest();
-            srcAddr   = copyObjOp->Source();
-            blockSize = copyObjOp->ClsTok();
-        }
-
-        noway_assert(destAddr->TypeGet() == TYP_BYREF || destAddr->TypeGet() == TYP_I_IMPL);
-        noway_assert(srcAddr->TypeGet() == TYP_BYREF || srcAddr->TypeGet() == TYP_I_IMPL);
+        JITDUMP("block assignment to morph:\n");
+        DISPTREE(asg);
 
-        unsigned blockWidth        = 0;
-        bool     blockWidthIsConst = false;
-
-        if (blockSize->IsCnsIntOrI())
+        if (dest->IsLocal())
         {
             blockWidthIsConst = true;
-            if (blockSize->IsIconHandle(GTF_ICON_CLASS_HDL))
+            destOnStack       = true;
+            if (dest->gtOper == GT_LCL_VAR)
             {
-                CORINFO_CLASS_HANDLE clsHnd = (CORINFO_CLASS_HANDLE)blockSize->gtIntConCommon.IconValue();
-                blockWidth                  = info.compCompHnd->getClassSize(clsHnd);
+                lclVarTree = dest->AsLclVarCommon();
+                destLclNum = lclVarTree->gtLclNum;
+                destLclVar = &lvaTable[destLclNum];
+                if (destLclVar->lvType == TYP_STRUCT)
+                {
+                    // It would be nice if lvExactSize always corresponded to the size of the struct,
+                    // but it does not for the spill temps.
```

Comment:
These aren't register spill temps - they're the temps that the importer uses when it needs to spill the stack. I'm trying to reconstruct where that happens, but I'm having some difficulty now. This is not something new; it's something that I found was true for existing code as well. I'll add a "TODO" if that's OK, as I think it may take some time to track this down again.

---
In reply to: [77418893](https://github.com/dotnet/coreclr/pull/7027#discussion_r77418893) [](ancestors = 77418893)